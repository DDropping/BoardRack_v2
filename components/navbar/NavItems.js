import React from "react";
import Link from "next/link";
import { useRouter } from "next/router";
import { useDispatch, useSelector } from "react-redux";
import { Dropdown } from "antd";
import { DownOutlined } from "@ant-design/icons";
import styled from "styled-components";

import { TOGGLE_LOGIN, TOGGLE_REGISTER } from "../../actions/types";
import Menu from "./AccountMenu";
import BadgeDot from "../badge/Dot";

const Ul = styled.ul`
  color: blue;
  display: inline-block;
  vertical-align: top;
  @media (max-width: ${(props) => props.theme.sm}) {
    display: none;
  }
`;

const Li = styled.li`
  transition: ${({ theme }) => theme.easeInOut};
  padding: 1rem 0.5rem 0.5rem 0.5rem;
  margin: 0 0.1rem;
  display: inline-block;
  cursor: pointer;
  font-weight: 600;
  ${({ active, theme }) =>
    active && `border-bottom: 2px solid ${theme.primaryBlue};`}
  &:hover {
    background-color: ${({ theme }) => theme.backgroundBlueMenu};
    border-bottom: 2px solid ${({ theme }) => theme.primaryBlue};
    a {
      color: ${({ theme }) => theme.primaryBlue};
    }
  }
`;

const NavItems = () => {
  const dispatch = useDispatch();
  const user = useSelector((state) => state.auth.user);
  const notifications = useSelector((state) => state.auth.notifications);
  const isAuth = useSelector((state) => state.auth.isAuthenticated);
  const isLogin = useSelector((state) => state.overlays.isLogin);
  const isRegister = useSelector((state) => state.overlays.isRegister);
  const router = useRouter();

  function isActive(route) {
    if (!router) {
      return false;
    }
    return route === router.pathname;
  }

  return (
    <Ul>
      {isAuth && (
        <Link href={"/createpost"}>
          <Li active={isActive("/createpost") && !isLogin && !isRegister}>
            <a className="create-post-link">Create Post</a>
          </Li>
        </Link>
      )}
      {!isAuth && (
        <Li onClick={() => dispatch({ type: TOGGLE_LOGIN, payload: true })}>
          <a className="create-post-link-disabled">Create Post</a>
        </Li>
      )}
      {isAuth && (
        <Dropdown overlay={<Menu />} overlayStyle={{ zIndex: 1000 }}>
          <Li active={isActive("/account") && !isLogin && !isRegister}>
            <a
              className="ant-dropdown-link"
              onClick={(e) => e.preventDefault()}
            >
              {user &&
                notifications.messages.length >
                  0 /* add more notification types here */ && (
                  <BadgeDot
                    size={12}
                    red
                    content={notifications.messages.length}
                  /> /* add more notification types here */
                )}{" "}
              {user ? user.username : "My Account"} <DownOutlined />
            </a>
          </Li>
        </Dropdown>
      )}
      {!isAuth && (
        <Li
          active={isLogin}
          onClick={() => dispatch({ type: TOGGLE_LOGIN, payload: true })}
        >
          <a className="login-link">Login</a>
        </Li>
      )}
      {!isAuth && (
        <Li
          active={isRegister}
          onClick={() => dispatch({ type: TOGGLE_REGISTER, payload: true })}
        >
          <a className="register-link">Register</a>
        </Li>
      )}
    </Ul>
  );
};

export default NavItems;
