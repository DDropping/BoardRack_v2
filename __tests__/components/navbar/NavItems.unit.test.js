import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";

import NavItems from "@components/navbar/NavItems";
import { initializeStore } from "@store";

let initialState, store, wrapper;

describe("NavItems when user is not authenticated", () => {
  beforeEach(() => {
    initialState = {};
    store = initializeStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <NavItems />
      </Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("has a create post link that will redirect to login", () => {
    const link = wrapper.find(".create-post-link-disabled");
    expect(link.length).toBe(1);
  });

  it("has a login link", () => {
    const link = wrapper.find(".login-link");
    expect(link.length).toBe(1);
  });

  it("has a register link", () => {
    const link = wrapper.find(".register-link");
    expect(link.length).toBe(1);
  });

  it("does not have a create post link", () => {
    const link = wrapper.find(".create-post-link");
    expect(link.length).toBe(0);
  });

  it("does not have a my account link", () => {
    const link = wrapper.find(".create-post-link");
    expect(link.length).toBe(0);
  });
});

describe("NavItems when user is authenticated", () => {
  beforeEach(() => {
    initialState = {
      auth: {
        token: "test_token",
        isAuthenticated: true,
        user: { username: "test_username", email: "test_email" },
      },
    };
    store = initializeStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <NavItems />
      </Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("has a create post link", () => {
    const link = wrapper.find(".create-post-link");
    expect(link.length).toBe(1);
  });

  it("has a my account link", () => {
    const link = wrapper.find(".ant-dropdown-link");
    expect(link.length).toBe(1);
  });

  it("has a my account link", () => {
    const link = wrapper.find(".ant-dropdown-link");
    expect(link.text()).toEqual("test_username ");
  });

  it("does not have a create post link that will redirect to login", () => {
    const link = wrapper.find(".create-post-link-disabled");
    expect(link.length).toBe(0);
  });

  it("does not have a login link", () => {
    const link = wrapper.find(".login-link");
    expect(link.length).toBe(0);
  });

  it("does not have a register link", () => {
    const link = wrapper.find(".register-link");
    expect(link.length).toBe(0);
  });
});

describe("NavItems when user is authenticated but loadUser() failed", () => {
  beforeEach(() => {
    initialState = {
      auth: {
        token: "test_token",
        isAuthenticated: true,
      },
    };
    store = initializeStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <NavItems />
      </Provider>
    );
  });

  it("has a my account link", () => {
    const link = wrapper.find(".ant-dropdown-link");
    expect(link.text()).toEqual("My Account ");
  });
});
