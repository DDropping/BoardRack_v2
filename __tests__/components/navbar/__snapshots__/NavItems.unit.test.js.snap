// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItems when user is authenticated but loadUser() failed matches snapshot 1`] = `
<ul
  className="sc-19qg9cj-0 eGTzxN"
>
  <li
    className="sc-19qg9cj-1 iTHAEh"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a
      className="create-post-link"
    >
      Create Post
    </a>
  </li>
  <li
    className="sc-19qg9cj-1 iTHAEh ant-dropdown-trigger"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <a
      className="ant-dropdown-link"
      onClick={[Function]}
    >
      My Account
       
      <span
        aria-label="down"
        className="anticon anticon-down"
        role="img"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="down"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
          />
        </svg>
      </span>
    </a>
  </li>
</ul>
`;

exports[`NavItems when user is authenticated matches snapshot 1`] = `
<ul
  className="sc-19qg9cj-0 eGTzxN"
>
  <li
    className="sc-19qg9cj-1 iTHAEh"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a
      className="create-post-link"
    >
      Create Post
    </a>
  </li>
  <li
    className="sc-19qg9cj-1 iTHAEh ant-dropdown-trigger"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <a
      className="ant-dropdown-link"
      onClick={[Function]}
    >
      test_username
       
      <span
        aria-label="down"
        className="anticon anticon-down"
        role="img"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="down"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
          />
        </svg>
      </span>
    </a>
  </li>
</ul>
`;

exports[`NavItems when user is not authenticated matches snapshot 1`] = `
<ul
  className="sc-19qg9cj-0 eGTzxN"
>
  <li
    className="sc-19qg9cj-1 iTHAEh"
    onClick={[Function]}
  >
    <a
      className="create-post-link-disabled"
    >
      Create Post
    </a>
  </li>
  <li
    className="sc-19qg9cj-1 iTHAEh"
    onClick={[Function]}
  >
    <a
      className="login-link"
    >
      Login
    </a>
  </li>
  <li
    className="sc-19qg9cj-1 iTHAEh"
    onClick={[Function]}
  >
    <a
      className="register-link"
    >
      Register
    </a>
  </li>
</ul>
`;
